src/osd/scrubber/scrub_machine.h

```diff
@@ -737,40 +774,95 @@ struct ReplicaActive : sc::state<ReplicaActive, ScrubMachine, ReplicaIdle>,
   // shortcuts:
   PG* m_pg;
   OSDService* m_osds;
-
-  /// a convenience internal result structure
-  struct ReservationAttemptRes {
-    MOSDScrubReserve::ReserveMsgOp op;	// GRANT or REJECT
-    std::string_view error_msg;
-    bool granted;
-  };
-
-  /// request a scrub resource from our local OSD
-  /// (after performing some checks)
-  ReservationAttemptRes get_remote_reservation();
-
-  void clear_reservation_by_remote_primary();
 };
 
 
-struct ReplicaIdle : sc::state<ReplicaIdle, ReplicaActive>, NamedSimply {
+struct ReplicaIdle : sc::state<
+			 ReplicaIdle,
+			 ReplicaActive,
+			 ReplicaUnreserved,
+			 sc::has_shallow_history>,
+		     NamedSimply {
   explicit ReplicaIdle(my_context ctx);
   ~ReplicaIdle() = default;
   void reset_ignored(const FullReset&);
+  using reactions = mpl::list<sc::in_state_reaction<
+      FullReset,
+      ReplicaIdle,
+      &ReplicaIdle::reset_ignored>>;
+};
+
+/*
+ * ReplicaUnreserved
+ *
+ * Possible events:
+ * - a reservation request from a legacy primary (i.e. a primary that does not
+ *   support queued reservations). We either deny or grant, transitioning to
+ *   ReplicaReserved directly.
+ * - a reservation request from a primary that supports queued reservations.
+ *   We transition to ProcessingReservation, and wait for the Reserver's
+ *   response.
+ * - (handled by our parent state) a chunk scrub request. We transition to
+ *   ReplicaActiveOp.
+ */
+struct ReplicaUnreserved : sc::state<ReplicaUnreserved, ReplicaIdle>,
+			   NamedSimply {
+  explicit ReplicaUnreserved(my_context ctx);
 
-  // note the execution of check_for_updates() when transitioning to
-  // ReplicaActiveOp/ReplicaWaitUpdates. That would trigger a ReplicaPushesUpd
-  // event, which will be handled by ReplicaWaitUpdates.
   using reactions = mpl::list<
-      sc::transition<
-	  StartReplica,
-	  ReplicaWaitUpdates,
-	  ReplicaActive,
-	  &ReplicaActive::check_for_updates>,
-      sc::in_state_reaction<
-	  FullReset,
-	  ReplicaIdle,
-	  &ReplicaIdle::reset_ignored>>;
+      sc::custom_reaction<ReplicaReserveReq>,
+      sc::custom_reaction<StartReplica>>;
+
+  sc::result react(const ReplicaReserveReq& eq);
+  sc::result react(const StartReplica& eq);
+};
+
+/**
+ * ProcessingReservation
+ *
+ * Possible events:
+ * - 'go ahead' from the asynch reserver. We send a GRANT message to the
+ *   primary & transition to ReplicaReserved.
+ * - 'cancel' from the primary. We clear our reservation state, and transition
+ *   back to ReplicaUnreserved.
+ * - a chunk request: best response is debateable. For now - we treat that
+ *   as a cancellation followed by the chunk request.
```

Comment:
I think we should handle the chunk response without affecting the reservation machinery at all.  It should be impossible, but it's not a problematic state for the replica so I don't think it's worth adding special code for.  It might be worth logging so that teuthology can fail the test, though.